runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/ProcessBundleDescriptors.java

```diff
@@ -0,0 +1,248 @@
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import org.apache.beam.model.fnexecution.v1.BeamFnApi;
+import org.apache.beam.model.fnexecution.v1.BeamFnApi.ProcessBundleDescriptor;
+import org.apache.beam.model.fnexecution.v1.BeamFnApi.ProcessBundleDescriptor.Builder;
+import org.apache.beam.model.fnexecution.v1.BeamFnApi.RemoteGrpcPort;
+import org.apache.beam.model.fnexecution.v1.BeamFnApi.Target;
+import org.apache.beam.model.pipeline.v1.Endpoints.ApiServiceDescriptor;
+import org.apache.beam.model.pipeline.v1.RunnerApi;
+import org.apache.beam.model.pipeline.v1.RunnerApi.Components;
+import org.apache.beam.model.pipeline.v1.RunnerApi.FunctionSpec;
+import org.apache.beam.model.pipeline.v1.RunnerApi.MessageWithComponents;
+import org.apache.beam.model.pipeline.v1.RunnerApi.PTransform;
+import org.apache.beam.model.pipeline.v1.RunnerApi.SdkFunctionSpec;
+import org.apache.beam.runners.core.construction.CoderTranslation;
+import org.apache.beam.runners.core.construction.graph.ExecutableStage;
+import org.apache.beam.runners.core.construction.graph.PipelineNode.PCollectionNode;
+import org.apache.beam.runners.core.construction.graph.PipelineNode.PTransformNode;
+import org.apache.beam.runners.fnexecution.control.SdkHarnessClient.RemoteInputDestination;
+import org.apache.beam.runners.fnexecution.graph.LengthPrefixUnknownCoders;
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.fn.data.RemoteGrpcPortRead;
+import org.apache.beam.sdk.fn.data.RemoteGrpcPortWrite;
+import org.apache.beam.sdk.util.WindowedValue;
+import org.apache.beam.sdk.util.WindowedValue.FullWindowedValueCoder;
+
+/** Utility methods for creating {@link ProcessBundleDescriptor} instances. */
+// TODO: Rename to ExecutableStages?
+public class ProcessBundleDescriptors {
+  public static SimpleProcessBundleDescriptor fromExecutableStage(
+      String id, ExecutableStage stage, Components components, ApiServiceDescriptor dataEndpoint) {
+    // Create with all of the processing transforms, and all of the components.
+    // TODO: Remove the unreachable subcomponents if the size of the descriptor matters.
+    ProcessBundleDescriptor.Builder bundleDescriptorBuilder =
+        ProcessBundleDescriptor.newBuilder()
+            .setId(id)
+            .putAllCoders(components.getCodersMap())
+            .putAllEnvironments(components.getEnvironmentsMap())
+            .putAllPcollections(components.getPcollectionsMap())
+            .putAllWindowingStrategies(components.getWindowingStrategiesMap())
+            .putAllTransforms(
+                stage
+                    .getTransforms()
+                    .stream()
+                    .collect(
+                        Collectors.toMap(PTransformNode::getId, PTransformNode::getTransform)));
+
+    RemoteInputDestination<WindowedValue<?>> inputDestination =
```

Comment:
Using RemoteInputDestination works but creates only half the story since the SdkHarnessClient also defines how to get output and you can't satisfy that right now since you don't have a gRPC service.

Maybe we should extract RemoteInputDestination as a different type which can be shared between this and the SdkHarnessClient so its not coupled there.